### ISSUE TITLE
[Feature request]: Add more method to Space such as flat_dim, flatten()

### ISSUE THREAD

**Issue by zuoxingdong** [openai/gym#1310]
I guess it might be very useful to add a few methods to the Space classes

e.g. 
```python
@property
def flat_dim(self):
        r"""Return a flattened dimension. """
        pass

@abstractmethod
def flatten(self, x):
        r"""Returns the flattened x. """
        pass
        
@abstractmethod
def unflatten(self, x):
        r"""Returns the unflattened x according to defined shape. """
        pass
```

This would be very useful for a clean and generic code for RL algorithm that handles both Discrete and Continuous control tasks, and for `flatten()/unflatten()` method could be useful when dealing neural networks over the observations. 

Another benefits is the methods are super easy to implement for nested spaces e.g. `Dict/Tuple`

What do you think @pzhokhov ?

**Comment by christopherhesse** [comment]
Could you give some examples of the input and output here?

**Comment by zuoxingdong** [comment]
@christopherhesse For example

```python 
fc1 = nn.Linear(env.observation_space.flat_dim, 128)
action_head = nn.Linear(128, env.action_space.flat_dim)

# e.g. Atari env
obs = env.reset()
obs = env.observation_space.flatten(obs)
features = fc1(obs)
action_features = action_head(features)
```
This would be very convenient for writing more generic code supporting both raw configuration and pixel observations, and both discrete and continuous control tasks. Also with more complex observation space (e.g. Nested tuple space), the user doesn't need to worry about handling MLP policy, the `flatten` method can deal with it automatically no matter which kind of space is.

**Comment by zuoxingdong** [comment]
@christopherhesse Some primary implementations are included in PR #1319

**Comment by wookayin** [comment]
TL;DR for who might find this on the Internet:

Use `gym.spaces.flatdim(env.observation_space)`, `gym.spaces.flatten(env.observation_space, ob)`, etc.

**Comment by audetto** [comment]
Is it possible that the implementation of these function is not correct?

IMHO, flat dim should return the sum and flatten should do something similar to a base transformation.

Reading the initial request, it seems they were added so that algorithms like q-learning do not need to worry about the shape, but they can treat the space as 1d. I am not sure how the implementation helps for this, nor what it is useful for.

https://github.com/openai/gym/issues/1830
